src/main/go/RelpConnection.go

```diff
@@ -0,0 +1,302 @@
+		success = true
+	}
+
+	if success {
+		relpConn.TearDown()
+	}
+
+	return success
+}
+
+// Commit commits the RELP batch to the server
+func (relpConn *RelpConnection) Commit(batch *RelpBatch) error {
+	if relpConn.state != STATE_OPEN {
+		panic("Can't commit, connection was in state other than OPEN.")
+	}
+
+	relpConn.state = STATE_COMMIT
+	err := relpConn.SendBatch(batch)
+	relpConn.state = STATE_OPEN
+	return err
+}
+
+// SendBatch sends the RELP frames to the server in the given batch.
+// The frames are sent asynchronously, and the server ACKs are checked after sending.
+func (relpConn *RelpConnection) SendBatch(batch *RelpBatch) error {
+	log.Printf("SendBatch.Entry> Work queue length: %v, Window.pending: %v : %v\n",
+		batch.GetWorkQueueLen(), len(relpConn.window.pending), relpConn.window.pending)
+	// send a batch of requests
+	for batch.GetWorkQueueLen() > 0 {
+		reqId := batch.PopWorkQueue()
+		relpRequest, err := batch.GetRequest(reqId)
+		if err != nil {
+			log.Fatalln("Could not get request from batch")
+		}
+
+		// relp Request-Response txId
+		// <txId is here> <command> <len> <data> NL
+		// make sure txId loops 1 - 999 999 999
+		if relpConn.txId >= 999_999_999 {
+			relpConn.txId = 1
+		} else {
+			relpConn.txId++
+		}
+		relpRequest.transactionId = relpConn.txId
+		log.Printf("SendBatch> Sending request\n%v %v %v '%v'\nfrom batch\n", relpRequest.transactionId, relpRequest.cmd,
+			relpRequest.dataLength, string(relpRequest.data))
+
+		relpConn.window.PutPending(relpConn.txId, reqId)
+		log.Println("SendBatch> Put pending: ", relpConn.txId, reqId)
+
+		go func() {
+			sendErr := relpConn.SendRelpRequest(relpRequest)
+			if sendErr != nil {
+				log.Printf("Error sending relp request: '%v'\n", err.Error())
+			}
+		}()
+	}
+
+	ackErr := relpConn.ReadAcks(batch)
+	if ackErr != nil {
+		// ACK timeout or other failure
+		return ackErr
+	}
+	return nil
+}
+
+// ReadAcks reads the ACKs from the given batch.
+func (relpConn *RelpConnection) ReadAcks(batch *RelpBatch) error {
+	log.Printf("ReadAcks.Entry>Reading ACKs for batchID: %v\n", batch.requestId)
+	var parser *RelpParser = nil
+	notComplete := relpConn.window.Size() > 0
+	var cn = *relpConn.connection
+
+	for notComplete {
+		tmp := make([]byte, 64)
+		readBytes := 0
+		for {
+			n, err := cn.Read(tmp)
+
+			err2 := cn.SetReadDeadline(time.Now().Add(relpConn.ackTimeoutDuration))
+			if err2 != nil {
+				return errors.New("error setting connection timeout")
+			}
+
+			if err != nil {
+				if errors.Is(err, os.ErrDeadlineExceeded) {
+					// reading timed out
+					return errors.New("reading ACK timed out")
+				} else if err == io.EOF {
+					return errors.New("encountered EOF in ACK (no connection?)")
+				} else {
+					// other error
+					return errors.New("unexpected error reading ACK")
+				}
+			}
+
+			// write and break if line break encountered
+			relpConn.preAllocRxBuffer.Write(tmp[0:n])
+			readBytes += n
+			if tmp[n-1] == '\n' {
```

Comment:
as relp is tlv protocol, the length must be used to read proper amount of bytes. this seems to be using terminating character which is not reliable.

Reply:
continues reading to buffer until RelpParser is complete
[7a7367f](https://github.com/teragrep/rlp_05/pull/1/commits/7a7367ff670bc0b3e99dc638a43fa7ede9f0284b)